File: 725-split-linked-list-in-parts/split-linked-list-in-parts.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def splitListToParts(self, head: Optional[ListNode], k: int) -> List[Optional[ListNode]]:
        length, cur = 0, head
        res = []

        while cur:
            cur = cur.next
            length += 1
        
        base = length // k
        rem = length % k

        cur = head

        for i in range(k):
            res.append(cur)
            for j in range(base - 1 + (1 if rem else 0)):
                if not cur: break
                cur = cur.next
            rem -= (1 if rem else 0)
            if cur:
                cur.next, cur = None, cur.next                
            
        return res

File: 725-split-linked-list-in-parts/split-linked-list-in-parts.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<ListNode*> splitListToParts(ListNode* head, int k) {
        int len = 0;
        vector<ListNode*> res;
        ListNode* cur = head;
        while (cur) {
            cur = cur->next;
            len++;
        }

        int base_len = len / k;
        int rem = len % k;
        cur = head;
        while (cur) {
            ListNode* temp = cur;
            int t = base_len;

            if (rem) {
                t++;
                rem--;
            }
            auto prev = cur;
            while (t--) {
                prev = cur;
                cur = cur -> next;
            }
            prev -> next = NULL;
            res.push_back(temp);
        }

        while (res.size() < k) {
            res.push_back(NULL);
        }

        return res;
    }
};